lib/core/proxy.c

```diff
@@ -130,8 +132,38 @@ static int req_requires_content_length(h2o_req_t *req)
     return is_put_or_post && h2o_find_header(&req->res.headers, H2O_TOKEN_TRANSFER_ENCODING, -1) == -1;
 }
 
+// copied from websocket.c FIXME move to somewhere?
+#include <openssl/sha.h>
+static void create_websocket_accept_key(char *dst, const char *client_key)
+{
+#define WS_GUID "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"
+    uint8_t sha1buf[20], key_src[60];
+
+    memcpy(key_src, client_key, 24);
+    memcpy(key_src + 24, WS_GUID, 36);
+    SHA1(key_src, sizeof(key_src), sha1buf);
+    h2o_base64_encode(dst, sha1buf, sizeof(sha1buf), 0);
+    dst[28] = '\0';
+#undef WS_GUID
+}
+
+static void create_websocket_key(char *dst)
+{
+    char rs[16];
+    int i;
+    for (i = 0; i != 16; i += 4) {
+        uint32_t r = h2o_rand();
+        rs[i] = (char)((r >> 24) & 0xFF);
+        rs[i + 1] = (char)((r >> 16) & 0xFF);
+        rs[i + 2] = (char)((r >> 8) & 0xFF);
+        rs[i + 3] = (char)(r & 0xFF);
+    }
+    h2o_base64_encode(dst, rs, 16, 0);
+    dst[24] = '\0';
+}
+
 static h2o_iovec_t build_request(h2o_req_t *req, int keepalive, int is_websocket_handshake, int use_proxy_protocol, int *te_chunked,
-                                 int *reprocess_if_too_early)
+                                 int *reprocess_if_too_early, char **websocket_key)
```

Comment:
Assuming that `websocket_key` is an "out" argument, I'd appreciate it if the function could always set a return value (i.e. IMO it should be set to NULL if there is no key).